In [1]:
# Import the necessary modules
import os
import numpy as np
import pandas as pd
from scipy.misc import *
from sklearn.metrics import accuracy_score
from matplotlib import pyplot as plt


# Keras and Tensorflow related imports
import tensorflow as tf
import keras

# To display Images
from IPython.display import display
from PIL import Image
%pylab inline

# Control Randomness
seed = 222
rng = np.random.RandomState(seed)

/anaconda/lib/python2.7/site-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


Populating the interactive namespace from numpy and matplotlib


/anaconda/lib/python2.7/site-packages/IPython/core/magics/pylab.py:161: UserWarning: pylab import has clobbered these variables: ['info', 'imsave', 'imshow', 'imread']
`%matplotlib` prevents importing * from pylab and numpy
  "\n`%matplotlib` prevents importing * from pylab and numpy"


#### Result
Populating the interactive namespace from numpy and matplotlib
Using TensorFlow backend.
/usr/local/lib/python3.6/site-packages/IPython/core/magics/pylab.py:160: UserWarning: pylab import has clobbered these variables: ['imread', 'imshow', 'source', 'imsave', 'who', 'info']
`%matplotlib` prevents importing * from pylab and numpy
  "\n`%matplotlib` prevents importing * from pylab and numpy"

In [2]:
# Directory Setup
# project_dir = os.getcwd()
# project_dir = '/Users/krishnamoorthypitchumani/PycharmProjects/Udacity Machine Learning/Udacity-Machine-Learning/projects/CapstoneProject'
project_dir = '/floyd/input/ageclassification'
train_img_dir = os.path.join(project_dir,'Data','train','Train')
test_img_dir = os.path.join(project_dir,'Data','test','Test')
train_data_dir = os.path.join(project_dir,'Data','train')
test_data_dir = os.path.join(project_dir,'Data','test')

# Get ImageNames
def get_image_name(n):
    yield str(n) + '.jpg'

# View random number of images
def get_images(n):
    for i in range(n):
        index = random.randint(0,10000)
        yield str(index) + '.jpg'
        
# Display Images
def display_images(n):
    image_list = [os.path.join(train_img_dir, image) for image in get_images(n)]
    print(image_list)

# Display Image from directory
def display(img_array):
    for img in img_array:
        print(img.shape)
        plt.imshow(img)
        plt.show()

In [81]:
# Load Input Data
train = pd.read_csv(os.path.join(train_data_dir, 'train.csv'))
test = pd.read_csv(os.path.join(test_data_dir, 'test.csv'))
train.head()

ID   Class
0    377.jpg  MIDDLE
1  17814.jpg   YOUNG
2  21283.jpg  MIDDLE
3  16496.jpg   YOUNG
4   4487.jpg  MIDDLE

In [64]:
# Data Preprocessing
from scipy.misc import imresize
from sklearn.preprocessing import LabelEncoder
import cv2

# Utility function to read images as arrays
def transform_image(img_path, mode = 'color', resize = False, size = 32):
    '''
    Default mode is : color(BGR) --> color(RGB)
    Other modes allowed are : 'grayscale' and 'include_opacity'
    '''
    if mode == 'grayscale':
        img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
#     elif mode == 'include_opacity':
#         img = cv2.imread(img_path, cv2.IMREAD_UNCHANGED)
    else:
        img = cv2.imread(img_path, cv2.IMREAD_COLOR)
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
  
    if resize == True:
        img = cv2.resize(img, (size, size))
    
    return img

# Resize_normalize images
def resize_images(dataset, img_dir):
    resized_array = []
    counter = 0
    for image_id in dataset.ID:
        img_path = os.path.join(img_dir, image_id)
        image = transform_image(img_path, resize=True, size = 64)
        image = image.astype('float32')
        resized_array.append(image)
        if counter == 19000:
            break
        counter += 1
    final_resized = np.stack(resized_array)
    return final_resized/255

# Encode the output class variable
def encode_feature(feature):
    lb = LabelEncoder()
    transformed_feature = lb.fit_transform(feature)
    transformed_feature = keras.utils.np_utils.to_categorical(transformed_feature)
    return transformed_feature

train_x = resize_images(train, train_img_dir)
test_x = resize_images(test, test_img_dir)
train_y = encode_feature(train['Class'])

#### Result

_________________________________________________________________
Layer (type)                 Output Shape              Param #   

=================================================================
flatten_2 (Flatten)          (None, 12288)             0         
_________________________________________________________________
dense_3 (Dense)              (None, 500)               6144500   
_________________________________________________________________
dense_4 (Dense)              (None, 3)                 1503      

=================================================================
Total params: 6,146,003
Trainable params: 6,146,003
Non-trainable params: 0
_________________________________________________________________
Train on 15200 samples, validate on 3801 samples

Epoch 1/20
15200/15200 [==============================] - 2s 133us/step - loss: 1.4581 - acc: 0.4446 - val_loss: 0.9266 - val_acc: 0.5454

Epoch 2/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.9205 - acc: 0.5569 - val_loss: 0.9037 - val_acc: 0.5556

Epoch 3/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8990 - acc: 0.5762 - val_loss: 0.8896 - val_acc: 0.5688

Epoch 4/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8950 - acc: 0.5689 - val_loss: 0.8833 - val_acc: 0.5712

Epoch 5/20
15200/15200 [==============================] - 1s 72us/step - loss: 0.8948 - acc: 0.5593 - val_loss: 0.8896 - val_acc: 0.5638

Epoch 6/20
15200/15200 [==============================] - 1s 71us/step - loss: 0.8724 - acc: 0.5891 - val_loss: 0.8628 - val_acc: 0.5962

Epoch 7/20
15200/15200 [==============================] - 1s 72us/step - loss: 0.8631 - acc: 0.5941 - val_loss: 0.8845 - val_acc: 0.5688

Epoch 8/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8862 - acc: 0.5680 - val_loss: 0.8573 - val_acc: 0.5933

Epoch 9/20
15200/15200 [==============================] - 1s 71us/step - loss: 0.8654 - acc: 0.5863 - val_loss: 0.8552 - val_acc: 0.5927

Epoch 10/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8643 - acc: 0.5850 - val_loss: 0.8766 - val_acc: 0.5733

Epoch 11/20
15200/15200 [==============================] - 1s 71us/step - loss: 0.8559 - acc: 0.5884 - val_loss: 0.8380 - val_acc: 0.6125

Epoch 12/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8387 - acc: 0.6126 - val_loss: 0.8475 - val_acc: 0.6033

Epoch 13/20
15200/15200 [==============================] - 1s 74us/step - loss: 0.8771 - acc: 0.5678 - val_loss: 0.8649 - val_acc: 0.5804

Epoch 14/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8516 - acc: 0.5915 - val_loss: 0.8326 - val_acc: 0.6143

Epoch 15/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8316 - acc: 0.6178 - val_loss: 0.8383 - val_acc: 0.6083

Epoch 16/20
15200/15200 [==============================] - 1s 74us/step - loss: 0.8696 - acc: 0.5758 - val_loss: 0.8580 - val_acc: 0.5827

Epoch 17/20
15200/15200 [==============================] - 1s 72us/step - loss: 0.8401 - acc: 0.6045 - val_loss: 0.8272 - val_acc: 0.6172

Epoch 18/20
15200/15200 [==============================] - 1s 73us/step - loss: 0.8277 - acc: 0.6159 - val_loss: 0.8344 - val_acc: 0.6096

Epoch 19/20
15200/15200 [==============================] - 1s 72us/step - loss: 0.8468 - acc: 0.5929 - val_loss: 0.8495 - val_acc: 0.5925

Epoch 20/20
15200/15200 [==============================] - 1s 74us/step - loss: 0.8328 - acc: 0.6040 - val_loss: 0.8226 - val_acc: 0.6175

<keras.callbacks.History at 0x7f55b92fada0>

In [36]:
train.Class[0:500]
train[501:510]

# train_x.shape
# valid_x[0][0][0]
# valid_y

ID   Class
501   5859.jpg   YOUNG
502  17169.jpg  MIDDLE
503   6691.jpg     OLD
504    537.jpg  MIDDLE
505  26003.jpg  MIDDLE
506   5069.jpg  MIDDLE
507   1330.jpg  MIDDLE
508  23418.jpg  MIDDLE
509    361.jpg     OLD

In [61]:
# Validation Data - This data can be used to test the model processing.
valid_x = resize_images(train[501:1001], train_img_dir)
valid_y = encode_feature(train['Class'][501:1001])
(valid_x, valid_y)


(array([[[[0.03921569, 0.03529412, 0.03529412],
          [0.03529412, 0.03529412, 0.03529412],
          [0.03137255, 0.03529412, 0.03921569],
          ...,
          [0.10980392, 0.07450981, 0.02745098],
          [0.07843138, 0.07843138, 0.03921569],
          [0.06666667, 0.07450981, 0.03921569]],
 
         [[0.04313726, 0.03921569, 0.02745098],
          [0.03529412, 0.03529412, 0.03529412],
          [0.03137255, 0.03921569, 0.03529412],
          ...,
          [0.11764706, 0.08235294, 0.03529412],
          [0.07843138, 0.07843138, 0.03921569],
          [0.06666667, 0.07450981, 0.03921569]],
 
         [[0.03921569, 0.03921569, 0.02352941],
          [0.03529412, 0.03529412, 0.02745098],
          [0.03137255, 0.03921569, 0.02745098],
          ...,
          [0.10588235, 0.07058824, 0.02745098],
          [0.07843138, 0.07843138, 0.03921569],
          [0.06666667, 0.07450981, 0.03921569]],
 
         ...,
 
         [[0.10196079, 0.10196079, 0.05490196],
          [0.09411

In [ ]:
# Benchmark Model
from keras.models import Sequential
from keras.layers import Flatten, Dense, InputLayer

# Variables & Parameters
input_units = (64, 64, 3)
hidden_units = 500
output_units = 3
epochs = 20
batch_size = 2000

# Model Creation
benchmark_model = Sequential([
                    InputLayer(input_shape=input_units),
                    Flatten(),
                    Dense(units=hidden_units, activation='relu'),
                    Dense(units=output_units, activation='softmax'),
                  ])

# Model Summary
benchmark_model.summary()

# Model Compilation
benchmark_model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

# Model Training & Validation
benchmark_model.fit(train_x, train_y[:10001], batch_size=batch_size,epochs=epochs,verbose=1,validation_split=0.2)

In [ ]:
# Make a prediction on the test set
test_predictions = model.predict(x_test)
test_predictions = np.round(test_predictions)# benchmark_model.evaluate(test_x, test_y[:10001], batch_size=batch_size,epochs=epochs,verbose=1)

# sample models

benchmark_model.predict_classes()

# model.predict_proba()
# evaluate the model
# scores = model.evaluate(X[test], Y[test], verbose=0)
# print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))
# cvscores.append(scores[1] * 100)

benchmark_model.metric_names

In [ ]:
# Model Implementation
from keras.layers import Conv2D, MaxPooling2D, GlobalAveragePooling2D
from keras.layers.normalization import BatchNormalization
from keras.layers import Dropout, Flatten, Dense
from keras.models import Sequential

# Variables & Parameters
batch_size = 500
num_classes = 3
epochs = 30
input_shape= (64, 64, 3)

# Model Creation
model = Sequential()
model.add(Conv2D(100, kernel_size=(2,2), activation='relu', padding='same', input_shape=input_shape ))
model.add(BatchNormalization())
model.add(Conv2D(200, kernel_size=(2,2), activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2,2), strides=(2,2)))
model.add(BatchNormalization())
model.add(Conv2D(300, kernel_size=(2,2), activation='relu', padding='same' ))
model.add(MaxPooling2D(pool_size=(2,2), strides=(2,2)))
model.add(BatchNormalization())
model.add(Flatten())
model.add(Dense(200, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(30, activation='relu'))
model.add(Dense(num_classes, activation='softmax'))

# Model Summary
model.summary()

# Model Compilation
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

# Model Training & Validation
model.fit(train_x, train_y[:10001], batch_size=batch_size,epochs=epochs,verbose=1,shuffle=True,validation_split=0.2)


_________________________________________________________________
Layer (type)                 Output Shape              Param #   

=================================================================

conv2d_1 (Conv2D)            (None, 64, 64, 100)       1300      
_________________________________________________________________
batch_normalization_1 (Batch (None, 64, 64, 100)       400       
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 64, 64, 200)       80200     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 32, 32, 200)       0         
_________________________________________________________________
batch_normalization_2 (Batch (None, 32, 32, 200)       800       
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 32, 32, 300)       240300    
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 16, 16, 300)       0         
_________________________________________________________________
batch_normalization_3 (Batch (None, 16, 16, 300)       1200      
_________________________________________________________________
flatten_3 (Flatten)          (None, 76800)             0         
_________________________________________________________________
dense_5 (Dense)              (None, 200)               15360200  
_________________________________________________________________
dropout_1 (Dropout)          (None, 200)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 30)                6030      
_________________________________________________________________
dense_7 (Dense)              (None, 3)                 93        

=================================================================

Total params: 15,690,523
Trainable params: 15,689,323
Non-trainable params: 1,200
_________________________________________________________________

Train on 15200 samples, validate on 3801 samples

Epoch 1/30
15200/15200 [==============================] - 78s 5ms/step - loss: 1.0668 - acc: 0.5914 - val_loss: 0.7805 - val_acc: 0.6435

Epoch 2/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.7265 - acc: 0.6763 - val_loss: 0.6796 - val_acc: 0.7093

Epoch 3/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.6702 - acc: 0.7046 - val_loss: 0.6793 - val_acc: 0.7096

Epoch 4/30
15200/15200 [==============================] - 69s 5ms/step - loss: 0.6254 - acc: 0.7312 - val_loss: 0.6607 - val_acc: 0.7219

Epoch 5/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.5800 - acc: 0.7564 - val_loss: 0.6170 - val_acc: 0.7459

Epoch 6/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.5428 - acc: 0.7760 - val_loss: 0.5940 - val_acc: 0.7580

Epoch 7/30
15200/15200 [==============================] - 68s 5ms/step - loss: 0.5015 - acc: 0.7963 - val_loss: 0.6198 - val_acc: 0.7440

Epoch 8/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.4661 - acc: 0.8140 - val_loss: 0.5746 - val_acc: 0.7622

Epoch 9/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.4375 - acc: 0.8249 - val_loss: 0.5753 - val_acc: 0.7574

Epoch 10/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.4039 - acc: 0.8425 - val_loss: 0.5820 - val_acc: 0.7632

Epoch 11/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.3775 - acc: 0.8561 - val_loss: 0.5629 - val_acc: 0.7748

Epoch 12/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.3515 - acc: 0.8653 - val_loss: 0.5582 - val_acc: 0.7748

Epoch 13/30
15200/15200 [==============================] - 68s 5ms/step - loss: 0.3252 - acc: 0.8797 - val_loss: 0.5867 - val_acc: 0.7669

Epoch 14/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.2998 - acc: 0.8880 - val_loss: 0.6146 - val_acc: 0.7543

Epoch 15/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.2726 - acc: 0.9035 - val_loss: 0.5536 - val_acc: 0.7824

Epoch 16/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.2598 - acc: 0.9080 - val_loss: 0.5524 - val_acc: 0.7816

Epoch 17/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.2300 - acc: 0.9232 - val_loss: 0.5337 - val_acc: 0.7922

Epoch 18/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.2150 - acc: 0.9303 - val_loss: 0.5788 - val_acc: 0.7764

Epoch 19/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.2032 - acc: 0.9338 - val_loss: 0.5364 - val_acc: 0.7916

Epoch 20/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1766 - acc: 0.9453 - val_loss: 0.5596 - val_acc: 0.7845

Epoch 21/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1655 - acc: 0.9495 - val_loss: 0.5438 - val_acc: 0.7895

Epoch 22/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1528 - acc: 0.9541 - val_loss: 0.5367 - val_acc: 0.7872

Epoch 23/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1434 - acc: 0.9572 - val_loss: 0.5527 - val_acc: 0.7932

Epoch 24/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1313 - acc: 0.9640 - val_loss: 0.5761 - val_acc: 0.7914

Epoch 25/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1224 - acc: 0.9688 - val_loss: 0.5619 - val_acc: 0.7998

Epoch 26/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1072 - acc: 0.9728 - val_loss: 0.5761 - val_acc: 0.7958

Epoch 27/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.1017 - acc: 0.9748 - val_loss: 0.5803 - val_acc: 0.7958

Epoch 28/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.0944 - acc: 0.9787 - val_loss: 0.5775 - val_acc: 0.7977

Epoch 29/30
15200/15200 [==============================] - 68s 4ms/step - loss: 0.0872 - acc: 0.9796 - val_loss: 0.5890 - val_acc: 0.7961

Epoch 30/30
15200/15200 [==============================] - 68s 5ms/step - loss: 0.0826 - acc: 0.9804 - val_loss: 0.6048 - val_acc: 0.7906

In [58]:
# Transfer Learning Implementation using VGG16
from keras.models import Sequential
from keras.models import Model
from keras.callbacks import ModelCheckpoint, LearningRateScheduler, EarlyStopping, ReduceLROnPlateau, TensorBoard
from keras import optimizers, models, activations, losses
from keras.layers import Convolution2D, Dense, Input, Flatten, Dropout, MaxPooling2D, BatchNormalization, \
GlobalAveragePooling2D, Concatenate
from keras.applications.vgg16 import VGG16
from keras.preprocessing import image
from keras.applications.vgg16 import preprocess_input

# Variables & Parameters
input_shape= (64, 64, 3)

# Import and make VGG16 model trainable = False and include_top = False to remove the top few layers
vgg16_base_model =  applications.VGG16(weights = 'imagenet',
                                            include_top = False,
                                            input_shape = input_shape)
vgg16_base_model.trainable = False

# Customise by adding Dense layers on top to fit our cause
custom_vgg16_model = Sequential()
custom_vgg16_model.add(vgg16_base_model)
custom_vgg16_model.add(GlobalAveragePooling2D())
custom_vgg16_model.add(Dropout(0.2))
custom_vgg16_model.add(Dense(3, activation='softmax'))

# Model Compilation
custom_vgg16_model.compile(loss='categorical_crossentropy', 
              optimizer=optimizers.SGD(lr=1e-4, 
                                       momentum=0.9),
              metrics=['accuracy'])

# Model Summary
custom_vgg16_model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
vgg16 (Model)                (None, 2, 2, 512)         14714688  
_________________________________________________________________
global_average_pooling2d_3 ( (None, 512)               0         
_________________________________________________________________
dropout_3 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 3)                 1539      
Total params: 14,716,227
Trainable params: 1,539
Non-trainable params: 14,714,688
_________________________________________________________________


### Result

Downloading data from https://github.com/fchollet/deep-learning-models/releases/download/v0.1/vgg16_weights_tf_dim_ordering_tf_kernels_notop.h5
58892288/58889256 [==============================] - 6s 0us/step
_________________________________________________________________

Layer (type)                 Output Shape              Param #   
vgg16 (Model)                (None, 2, 2, 512)         14714688  
_________________________________________________________________
global_average_pooling2d_1 ( (None, 512)               0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 3)                 1539      

Total params: 14,716,227
Trainable params: 1,539
Non-trainable params: 14,714,688

_________________________________________________________________


In [5]:
weights_file = 'top_weights.hd5'
checkpoint = ModelCheckpoint(weights_file, monitor='acc', verbose=1, save_best_only=True, mode='max')
callbacks = [checkpoint]
batch_size = 50
epochs = 10

# Model Training
custom_vgg16_model.fit(train_x, train_y[:19001], batch_size=batch_size,epochs=epochs,verbose=1,shuffle=True,validation_split=0.2)


Train on 15200 samples, validate on 3801 samples
Epoch 1/20
15200/15200 [==============================] - 19s 1ms/step - loss: 0.9302 - acc: 0.5786 - val_loss: 0.8996 - val_acc: 0.5925
Epoch 2/20
15200/15200 [==============================] - 15s 962us/step - loss: 0.9300 - acc: 0.5782 - val_loss: 0.8984 - val_acc: 0.5975
Epoch 3/20
15200/15200 [==============================] - 15s 961us/step - loss: 0.9259 - acc: 0.5834 - val_loss: 0.8973 - val_acc: 0.5985
Epoch 4/20
15200/15200 [==============================] - 15s 964us/step - loss: 0.9251 - acc: 0.5823 - val_loss: 0.8963 - val_acc: 0.5967
Epoch 5/20
15200/15200 [==============================] - 15s 962us/step - loss: 0.9268 - acc: 0.5802 - val_loss: 0.8953 - val_acc: 0.5985
Epoch 6/20
15200/15200 [==============================] - 15s 968us/step - loss: 0.9247 - acc: 0.5826 - val_loss: 0.8943 - val_acc: 0.5993
Epoch 7/20
15200/15200 [==============================] - 15s 966us/step - loss: 0.9212 - acc: 0.5839 - val_loss: 0.8934 - val_acc: 0.5988
Epoch 8/20
15200/15200 [==============================] - 15s 968us/step - loss: 0.9229 - acc: 0.5805 - val_loss: 0.8926 - val_acc: 0.5996
Epoch 9/20
15200/15200 [==============================] - 15s 966us/step - loss: 0.9262 - acc: 0.5803 - val_loss: 0.8918 - val_acc: 0.6004
Epoch 10/20
15200/15200 [==============================] - 15s 971us/step - loss: 0.9214 - acc: 0.5790 - val_loss: 0.8910 - val_acc: 0.6004
Epoch 11/20
15200/15200 [==============================] - 15s 968us/step - loss: 0.9203 - acc: 0.5831 - val_loss: 0.8902 - val_acc: 0.5996
Epoch 12/20
15200/15200 [==============================] - 15s 969us/step - loss: 0.9173 - acc: 0.5853 - val_loss: 0.8895 - val_acc: 0.5998
Epoch 13/20
15200/15200 [==============================] - 15s 966us/step - loss: 0.9206 - acc: 0.5835 - val_loss: 0.8888 - val_acc: 0.6006
Epoch 14/20
15200/15200 [==============================] - 15s 969us/step - loss: 0.9182 - acc: 0.5839 - val_loss: 0.8881 - val_acc: 0.6009
Epoch 15/20
15200/15200 [==============================] - 15s 967us/step - loss: 0.9150 - acc: 0.5861 - val_loss: 0.8873 - val_acc: 0.6022
Epoch 16/20
15200/15200 [==============================] - 15s 968us/step - loss: 0.9156 - acc: 0.5862 - val_loss: 0.8868 - val_acc: 0.6025
Epoch 17/20
15200/15200 [==============================] - 15s 965us/step - loss: 0.9116 - acc: 0.5892 - val_loss: 0.8861 - val_acc: 0.6014
Epoch 18/20
15200/15200 [==============================] - 15s 966us/step - loss: 0.9131 - acc: 0.5849 - val_loss: 0.8855 - val_acc: 0.6030
Epoch 19/20
15200/15200 [==============================] - 15s 967us/step - loss: 0.9133 - acc: 0.5885 - val_loss: 0.8848 - val_acc: 0.6022
Epoch 20/20
15200/15200 [==============================] - 15s 968us/step - loss: 0.9138 - acc: 0.5839 - val_loss: 0.8842 - val_acc: 0.6022